### ISSUE TITLE
Boston_Housing 모델학습 관련하여 질문드립니다.

### ISSUE THREAD

**Issue by sinclair2323** [hsnam95/class2019Spring#9]
Boston_Housing 데이터를 사용해 모델을 학습시킬 때, 은닉층의 수와 epoch수를 늘리면 train과정에서의 loss는 많이 줄어들지만, test data로 모델을 evaluate시켰을 때 loss가 오히려 높아지는 오버피팅 문제가 발생한다는 것을 알게 되었습니다.

오버피팅 문제를 해결하기 위해 다음과 같이 모델을 수정해보았습니다.
- 13개의 feature들 중 'chas'가 집값과 큰 상관을 보이는 것 같지 않아 입력에서 제외시켰습니다. (모델의 복잡성을 줄이기 위해 feature의 개수를 줄임)
- 오버피팅이 일어나기 전에 학습을 조기종료시키기 위하여 epoch을 1000에서 600으로 줄였습니다.
- relu 활성화 함수가 부여된 3개의 은닉층을 만들고, 각각 10, 15, 10개의 노드로 구성했습니다.
- epoch을 줄였기 때문에 짧은 학습시간에도 가중치를 충분히 갱신하기 위해서 학습률을 0.001에서 0.005로 높였습니다.
 
![image](https://user-images.githubusercontent.com/45940013/57857907-07206080-782b-11e9-8f41-b9c58997f41d.png)

그 결과 train loss, test loss가 (25, 30)에서 (10, 21)로 감소했습니다. 

많은 시행착오를 통해 제가 느낀 것은
- 위에서 언급한 입력 feature의 개수, epoch, 은닉층의 개수, 노드 수, 학습률 등의 하이퍼파라미터 중 어떤 하나를 수정하기만해도 loss값이 높아지기도해 다른 하이퍼파라미터들의 수정이 필요했다는 점
- 동일한 코드임에도 불구하고 학습을 새로 할 때마다 (가중치 초기화 시 랜덤을 이용하기 때문에) loss값이 크게(10정도?) 달라질 수 있다는 점
입니다.

**이러한 이유로, 제가 조정을 통해 만든 모델의 성능 향상 또는 저하가 1) 하이퍼파라미터를 조정해서인지, 2) 매 학습마다 다른 값이 나오는 학습 자체의 특성 때문인지 구별하기 어렵습니다.** 
**이 둘을 구분하는 기준이 없을까요?**
**또 loss값이 어느 정도로 줄어들어야 성능향상의 원인이 2)가 아니라고 판단할 수 있나요?**

**Comment by cjw414** [comment]
1)은 복합적인 문제라 확답을 못 드리지만, 2)에 대해서는 어느 정도 도움을 드릴 수 있을 거 같습니다.
케라스를 불러오기  전에 난수를 지정해줘서 모델의 weight값을 (크기가 동일하다면) 일정하게 나올 수 있는 거로 알고 있습니다. 
https://github.com/keras-team/keras/issues/2743
에서 볼 수 있듯이, Sequential모델을 불러오기 전에 np.seed(1004)처럼 난수를 지정해주고 다양한 파라미터들을 사용하면 더 좋은 결과를 내실 수 있을 겁니다.

import numpy as np
np.random.seed(1337)  # for reproducibility
from tensorflow.keras.models import Sequential

**Comment by sinclair2323** [comment]
감사합니다. 많은 도움이 되었습니다.
